sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3468,6 +3448,52 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Restricts the scope of resolving some expressions.
+   */
+  object ScopeExpressions extends Rule[LogicalPlan] {
+    private def scopeOrder(scope: Seq[Attribute])(sortOrder: SortOrder): SortOrder = {
+      sortOrder match {
+        case so if so.child.isInstanceOf[ScopedExpression] => so
+        case so => so.copy(child = ScopedExpression(so.child, scope))
```

Comment:
`SortOrder` has `sameOrderExpressions`, shall we scope them as well, or assert they are empty here?

Reply:
Added in 894368e1432.